internal/resolution/datasource/npmrc_test.go

```diff
@@ -154,3 +156,141 @@ func TestNpmrcRegistryOverriding(t *testing.T) {
 	t.Setenv("NPM_CONFIG_REGISTRY", "https://environ.registry.com")
 	check(t, npmrcFiles)
 }
+
+func TestNpmRegistryAuthOpts(t *testing.T) {
+	t.Parallel()
+	b64enc := func(s string) string {
+		t.Helper()
+		return base64.StdEncoding.EncodeToString([]byte(s))
+	}
+	tests := []struct {
+		name       string
+		opts       datasource.NpmRegistryAuthOpts
+		requestURL string
+		wantAuth   string
+	}{
+		// Auth tests adapted from npm-registry-fetch
+		// https://github.com/npm/npm-registry-fetch/blob/237d33b45396caa00add61e0549cf09fbf9deb4f/test/auth.js
+		{
+			name: "basic auth",
+			opts: datasource.NpmRegistryAuthOpts{
+				"//my.custom.registry/here/:username":  "user",
+				"//my.custom.registry/here/:_password": b64enc("pass"),
+			},
+			requestURL: "https://my.custom.registry/here/",
+			wantAuth:   "Basic " + b64enc("user:pass"),
+		},
+		{
+			name: "token auth",
+			opts: datasource.NpmRegistryAuthOpts{
+				"//my.custom.registry/here/:_authToken": "c0ffee",
+				"//my.custom.registry/here/:token":      "nope",
+				"//my.custom.registry/:_authToken":      "c0ffee",
```

Comment:
Should we set different tokens for the two paths here (/here vs / )? To make sure the right one is selected?

Reply:
Will do.
For some reason npm-registry-fetch has these the same in [this test](https://github.com/npm/npm-registry-fetch/blob/237d33b45396caa00add61e0549cf09fbf9deb4f/test/auth.js#L55-L63):